src/AstSemanticChecker.cpp

```diff
@@ -745,16 +745,73 @@ void AstSemanticChecker::checkRules(ErrorReport& report, const TypeEnvironment&
 
 // ----- types --------
 
+// check if a union contains a number primitive
+static bool unionContainsNumber(const AstProgram& program, const AstUnionType& type) {
+    for (const AstTypeIdentifier& sub : type.getTypes()) {
+        if (sub == "number") {
+            return true;
+        }
+        const AstType* subt = program.getType(sub);
+        if (const auto* u = dynamic_cast<const AstUnionType*>(subt)) {
+            if (unionContainsNumber(program, *u)) {
```

Comment:
This could just be `return unionContainsNumber(program, *u);`, and similarly for the others below.

Reply:
Not quite, since this is in a loop and some later iteration might return true, we don't want to prematurely return false

Reply:
Hopefully the comments added in e7403c697c16c269f500fcee3a5ccfa82a735d63 make the intended behavior more clear

Reply:
Good point.